### ISSUE TITLE
New transactions don't correctly set timezone

### ISSUE THREAD

**Issue by mehalter** [dreautall/waterfly-iii#246]
I have recently noticed a strange behavior when creating transactions on my phone when in a different time zone as my firefly server. It seems that waterfly sets the date/time in the transaction to the exact value of my phone but doesn't do any translation to the timezone of the server. 

For example: I'm currently in a time zone 13 hours ahead of my server and all transactions I make in waterfly are created 13 hours in the future.

I would expect to set the transaction date/time to the local time on my phone but then it be translated  and set to the correct time on the server based on its time zone.

**Comment by mehalter** [comment]
Yeah maybe this is something that should actually be discussed upstream. Something like the timezone displayed in a web server I think also should have the option to display in the current device timezone. This is information fully available in a progressive web application and should be utilized to provide the information in a unified way with the device to provide a clear user interface and minimize confusion.

Another feature I think makes sense to be upstream is to allow the setting of the timezone for a new transaction time rather than assuming the server timezone. Then the answer becomes clear in the app which is: have an option to follow the device timezone or display in the server timezone, then you can simply provide the timezone information in the API call to get the correct storage behavior. This also could be "just follow firefly web, and if you have auto detect timezone" option as described above set then you know. Then you can get the correct behavior in the app while staying aligned with the web server.

TL;DR is follow the firefly web server because that's the most clear solution from the app's perspective, and we hope that upstream the web server adds support for auto detecting timezone of the browser 😂

What do you think?

**Comment by mehalter** [comment]
Yeah hopefully I can open an issue upstream and get support added for proper timezone strings in the API

**Comment by mehalter** [comment]
I opened up a feature request upstream to get full timezone support added to the UI and API and hopefully compliance with something like ISO86001 standard

**Comment by JC5** [comment]
The problem I had in the past with this is database support. It's terrible. I support this as much as the next guy but don't hold your breath. I'm not going to break hundreds of installations for this. I did that already when I tried this before. 

Firefly iii assumes everything is in the servers timezone. I don't mind making an endpoint to show you that timezone. But you'll have to do the math yourself for the time being. 

For now I'm solving this problem by ignoring it.

**Comment by mehalter** [comment]
Thanks for the response @JC5. I definitely agree that changing how the database stores things is not the best approach. I think it can mainly be solved on the rendering side where we can submit transactions with timezone information and the API endpoint translates it to the server timezone and stores it exactly as it is currently. No need to change the database in any way. Then on the rendering side you can simply do the opposite. Detect the timezone of the device in browser and then calculate the offset based on the server timezone

**Comment by JC5** [comment]
> where we can submit transactions with timezone information and the API endpoint translates it to the server timezone and stores it exactly as it is currently

This already happens. It's easy to test this. Submitting `"date": "2024-01-01T07:13:51+04:00",` will be corrected by Firefly III.

> Everything you enter, regardless of your own timezone, will be treated as "server timezone" (and, if you change the server timezone, everything will stay at the same time of the day, just with a new timezone!).

For the Firefly III UI, this is also true. Because most browsers, when in any tracker-block-mode, pretend you are in UTC to protect your privacy. So the UI doesn't submit any timezone information.

**Comment by mehalter** [comment]
Ah if the API does support timezone inclusive datetime strings, then maybe it is a good idea for us to do our own timezone calculation in waterfly @dreautall . The mobile app definitely provides a more portable and integrated experience with the users device so it would be nice for things to align with the location of the device and the system time rather than it being separate. Then we can do all of our transaction insertions and things with timezone inclusive strings to make sure that firefly inserts with the correct datetime

**Comment by mehalter** [comment]
Since firefly is very much not intending to add timezone support upstream I do think trying to do Case 2 in the app would be a good idea. We can even have a toggle for "server timezone" or "local timezone" and the implementation will capture the ability to handle both since the "server timezone" would just be hard coding what the device timezone looks like. The main reason here being that waterfly, being a phone application, definitely provides a more mobile interface into firefly and is meant to be "on the go" and integrated with my phone device. So it would make sense in my mind for the app to integrate with the devices system time.

I do think that it would be a good idea to somehow query other users regarding this just to get other opinions. I can see pros on each side which leads me to think that an implementation that can easily support both cases with a setting toggle seems the most ideal.

**Comment by mehalter** [comment]
Wow that's awesome! Thanks so much @dreautall ! I sent you an email with my Google account mail address for the alpha.

**Comment by mehalter** [comment]
I side loaded from the GitHub releases and it seems to work great when I set the timezone to local timezone! I can view the transactions in the correct local time and correctly submit new transactions as well. I will do some more testing over the next few days of editing and things as well.

One issue I found is with the "server timezone" setting. When I go it make a new transaction it puts in the direct time on my phone rather than converting it to the server timezone first. So it gets inserted as if the current device time is in the server timezone (in my case, 13 hours ahead), but other than that everything on that side looks great as well!

**Comment by mehalter** [comment]
I think I would also recommend changing the toggle to be something  like "use system time" and off follow the server timezone and it be the default. Rather than default is on and it says "use server time"

**Comment by mehalter** [comment]
Ah I see what you mean by the "intended behavior". I would say if you are replicating the web server it seems to always insert transactions by default at 12a rather than at the devices current time. I just think it definitely seems like a bug when I say "new transaction" and "save" it inserts at the strictly incorrect time. I would think that at the very least it should convert the time to the server time in the transaction since it should fill in the transaction with the server time if it's actually doing that. In this case it's strictly not filling it in with the server time if that makes sense. Idk, maybe other people will have a stronger opinion but it does seem like a bug when the app fills in a specific time and it is wrong in all interpretations of it

**Comment by mehalter** [comment]
Basically I think having a server time option as it is currently is good but for new transactions it should actually insert the time on the server which it doesn't do. It inserts the time on the device and then says "this is the server time zone"

**Comment by mehalter** [comment]
@dreautall I just mean the time that gets auto filled by the app. I agree if I put in 1p it should insert as that. The insertion is working perfectly. It's the auto filling of the app. I'm *not* putting 1p, the app is filling in automatically the incorrect server time. It's filling in my device time when I'm not saying "follow the device time". I'm saying follow the server time, so it should insert the current time on the server automatically in the form.

**Comment by mehalter** [comment]
Basically let me describe a workflow. I'm currently in a timezone 12 hours ahead of my home timezone. I have the app set to "server time" and I go to waterfly and say "create new transaction" and in my local timezone it's 1p. I fill in the accounts and the cost and it is filled in that it's 1p. I click submit. I go to firefly web and it shows that the transaction has *not* happened yet. It will happen in 12 hours because on the server it is 1a.

I think what should happen is I click "new transaction" and it should say that it's 1a automatically because that is when the transaction actually happened. If the user wants it to be local then they should say "follow device time". Or they can manually set 1p and it will insert as 1p as the server time.


I think the important part is that transactions happen linearly in time in relation to each other. And when they truly, actually happened.

**Comment by mehalter** [comment]
Actually, following the same workflow @dreautall its more confusing just in the app with the timezone set to "server timezone". Because the transaction is automatically set to the future because it interprets the "device time" as "server time". If you follow the previous workflow and create a new transaction at 1p, it automatically fills in 1p. Then you refresh the app and you won't see your transaction because it hasn't happened yet since it's only 1a on the server. You will only see it once the actual time has passed or you say "future transactions" filter. This causes some extreme confusion when you make a transaction, refresh and it's not shown. It makes you think there was an error and a user might think the transaction wasn't made and then tries again. This was the issue I ran into initially which made me open this issue in the first place.

**Comment by mehalter** [comment]
Thanks so much! I'll happily continue testing as well once you get the fix added 😄 these are definitely some specific situations I haven't run into before even in the 5+ years of using firefly. My guess is not many people run into this type of issue very often 😅

I really appreciate all of your help with this issue and this app implementation in general! I've been waiting a long time for a high quality app to be made to replace just using the mobile firefly-web site. I have tried many over the years and none has even come close to where this app is!

**Comment by mehalter** [comment]
Just downloaded the latest build with the most recent commits. It still doesn't follow actual timezone stuff. Like it's still putting my device time rather than the time on the server. So it's still putting transactions in the future. Like when I make a transaction in "server time" it should be defaulting the transaction time to the *actual* time on the server at that moment. Not the time on my phone. This will cause some extreme inconsistencies in the linear passing of time and the ordering of transactions during travel taking place. It also messes up graphs since the last transaction in one time zone will be arbitrarily spaced with the first transaction in another time zone. So the idea of time starts breaking down. I think what you are saying in previous messages is it seems weird if you have "lunch" at 1a in your home time zone, but in reality you did have lunch then so it seems weird to not make the default behavior to apply the correct time. The current implementation of "use server time" is actually a weird mixture of the device time and the server timezone.

I am not using the server time setting anyway, so it seems to at least be working the way I want it in the settings I use. My guess is the next person who does large international travel will come to similar confusions when their transactions start happening at seemingly random/arbitrary times.

Thanks again for your help on this and getting at least the local device timezone integration in there!
